activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1723,8 +1723,8 @@ def build_join_dependencies
         )
       end
 
-      def assert_mutability!
```

Comment:
@fxn would you consider keeping this method name (or aliasing) for a while? It's breaking some downstream apps, eg. https://github.com/salsify/goldiloader/issues/150. I know it's not public API :(

Reply:
I don't want to break applications, but `assert_mutability!` was not only private API, it was `private` Ruby visibility.

![CleanShot 2024-07-16 at 08 11 23@2x](https://github.com/user-attachments/assets/c48cebc6-ce58-41da-8512-0cd055b8d9ab)

Doing deprecation cycles for public API is to be expected.

Public methods that are not public API ... in general they are supposed to be private, by default you don't deprecate, but in some occasions maybe you could do an exception if truly justified.

But `private` Ruby visibility is the extreme, the contract is that you can freely change this. If we should be careful for refactoring private APIs, Rails development would come to a halt. And that does not change in the presence of someone depending on that private method.

I don't have anything against using private APIs as long as you guard them and take full responsibility for using them.

In this case, Goldiloader should release a fix. I see they have a good history of saying what works with what. Which is aligned with my comment above. The project monkey patches, and knows its monkey patches depend on the Rails version, and updates them.

I'd consider an alias only in the unlikely case there is no response from the project.